### ISSUE TITLE
[RFC] RLHF follow-ups

### ISSUE THREAD

**Issue by SalmanMohammadi** [pytorch/torchtune#1395]
There are several optimizations to our PPO recipe which could help push it closer to SOTA in terms of performance. There are also several pieces of documentation we could offer alongside this recipe to increase visibility and improve accessibility. These are non-comprehensive and not all required.

### Documentation 
---
- Recipe documentation page which sufficiently explains how to use the recipe, including:
    - model and dataset requirements
    - explanation of all recipe parameters and algorithm hyperparameters 
    - artifacts from a reference baseline training run 
    - (see #1230).
- Recipe tutorial guiding a user through the PPO fine-tuning process. This tutorial will only provide a high-level overview of the algorithm, and cover: 
  - how to identify a suitable reward model and dataset for a given task
  - downloading models and guidance on early experimentation (start small and fast)
  - tuning parameters/things to look out for/how you know it's working
  - evaluating trained models (qualtiative analysis w/ generations, quantitative analysis through an appropriate eleuther eval)
  -  uploading to hub?
- E2E RLHF workflow in torchtune - SFT -> Reward modelling -> PPO
 - Recipe deepdive covering:
   - Explanation of relevant papers and priorart
   - Implementation details a la [1](https://iclr-blogposts.github.io/2024/blog/the-n-implementation-details-of-rlhf-with-ppo/), [2](https://arxiv.org/abs/2403.17031)

### Optimizations

Rough benchmarks from [deepspeed](https://github.com/microsoft/DeepSpeed/blob/master/blogs/deepspeed-chat/README.md#capability-recap)

I think the results from this page all use LoRA. Nonetheless, it's one of the only sources of compute useage for a modern RLHF implementation.

<img width="709" alt="image" src="https://github.com/user-attachments/assets/e653e8e9-de52-4ea5-9157-fb6d4aa0eb9c">

*It's unclear what size of reward model is used here. Throughout the blogpost they use reward model sizes << policy model sizes.

<img width="1299" alt="image" src="https://github.com/user-attachments/assets/042db35c-17ff-4fa8-a37c-f046f9bfd4a0">

They also state:
> For now, we suggest that users use "Total-GPU-Memory-in-GB / 6" as the upper parameter bound in billions for the sum of the actor model and critical model, for safety. Nevertheless, users are welcome to try the real limit.

Which gives 13.B for the combined memory of both actor + critic model on a single A100 80GB.

---
### **Compile issues** @ebsmothers 

- [ ] Fix compile for recipe? (**blocker:** currently hitting recompile cache limit due to swapping between inference-training modes - plus many more)

---

- [x] #1402
   - Managed to solve initial graph breaks due to rng object, by hiding away sampling logic into a util and outside of the compiled fn
   - New blocker: no more graph breaks and an initial recompile, but very slow 
   - Fixed above issues. Refactoring and cleaning up - seeing ~20% speedups with compile.

- [ ] Enable compile for trajectory generation step?
- [ ] Enable compile for loss step?
- ?? how else can we make inference go fast?
- [ ] Reference +/ reward model offload to CPU @ebsmothers 
- [ ] Optimizer offload to CPU (#1351) (to benchmark once it lands)
- (From deepspeed link above) - granted these aren't strictly performance opt:
>  Exponential Moving Average (EMA) collection, where an EMA based checkpoint can be chosen for the final evaluation.
    Mixture Training, which mixes the pretraining objective (i.e., the next word prediction) with the PPO objective to prevent regression performance on public benchmarks like SQuAD2.0.
- [ ] Add LoRA PPO
- [ ] MARL (https://huggingface.co/docs/trl/multi_adapter_rl) - single base model, multi-adapter PPO training

cc @kartikayk

**Comment by SalmanMohammadi** [comment]
Moving compiling generation out to a separate comment.

So compiled is significantly slower.
```bash
(tune) salman@combuter:~/torchtune$ python generate_test.py rlhf_compile
Warmup run!
Time for warmup: 23.83 sec
V0823 13:04:27.827000 148928 torch/_dynamo/guards.py:2688] [0/1] [__recompiles] Recompiling function generate_next_token_with_logits in /home/salman/torchtune/torchtune/modules/rlhf/_generation.py:37
V0823 13:04:27.827000 148928 torch/_dynamo/guards.py:2688] [0/1] [__recompiles]     triggered by the following guard failure(s):
V0823 13:04:27.827000 148928 torch/_dynamo/guards.py:2688] [0/1] [__recompiles]     - 0/0: tensor 'L['input_pos']' stride mismatch at index 0. expected 130, actual 192
Output:
 
...
Time for inference: 27.96 sec total, 21.78 tokens/sec
Bandwidth achieved: 112.15 GB/s
Memory used: 5.90 GB

(tune) salman@combuter:~/torchtune$ python generate_test.py rlhf
Output:
 
...
Time for inference: 5.49 sec total, 110.99 tokens/sec
Bandwidth achieved: 571.61 GB/s
Memory used: 5.21 GB

 ```

---

Now, if I set the backend to cudagraphs and log:
```python
I0823 14:22:20.582000 276899 torch/_inductor/cudagraph_trees.py:364] [__cudagraphs] recording cudagraph tree for graph without symints
V0823 14:22:20.726000 276899 torch/_inductor/cudagraph_trees.py:2099] [__cudagraphs] Running warmup of function 0
Time for warmup: 11.63 sec
V0823 14:22:20.848000 276899 torch/_dynamo/guards.py:2688] [0/1] [__recompiles] Recompiling function generate_next_token_with_logits in /home/salman/torchtune/torchtune/modules/rlhf/_generation.py:37
V0823 14:22:20.848000 276899 torch/_dynamo/guards.py:2688] [0/1] [__recompiles]     triggered by the following guard failure(s):
V0823 14:22:20.848000 276899 torch/_dynamo/guards.py:2688] [0/1] [__recompiles]     - 0/0: tensor 'L['input_pos']' stride mismatch at index 0. expected 356, actual 610
I0823 14:22:30.863000 276899 torch/_inductor/cudagraph_trees.py:364] [__cudagraphs] recording cudagraph tree for graph without symints
V0823 14:22:30.865000 276899 torch/_inductor/cudagraph_trees.py:2099] [__cudagraphs] Running warmup of function 1
V0823 14:22:30.955000 276899 torch/_inductor/cudagraph_trees.py:2101] [__cudagraphs] Running eager of function 1 because ancestor needed to warm up
/home/salman/.pyenv/versions/3.11.9/envs/tune/lib/python3.11/site-packages/torch/_inductor/cudagraph_trees.py:2282: UserWarning: Unable to hit fast path of CUDAGraphs because of pending, uninvoked backwards. Consider running with torch.no_grad() or using torch.compiler.cudagraph_mark_step_begin() before each model invocation
  warnings.warn(
V0823 14:22:31.048000 276899 torch/_inductor/cudagraph_trees.py:2101] [__cudagraphs] Running eager of function 1 because ancestor needed to warm up
V0823 14:22:31.141000 276899 torch/_inductor/cudagraph_trees.py:2101] [__cudagraphs] Running eager of function 1 because ancestor needed to warm up
V0823 14:22:31.229000 276899 torch/_inductor/cudagraph_trees.py:2101] [__cudagraphs] Running eager of function 1 because ancestor needed to warm up
... continues on max_generated_tokens times
```

**Comment by gau-nernst** [comment]
Not sure if it will be useful for you, but there are 8-bit and 4-bit AdamW in torchao https://github.com/pytorch/ao/tree/main/torchao/prototype/low_bit_optim. Both support FSDP1/2.
8-bit AdamW should match bnb and 4-bit version should match lpmm exactly. They are included in torchao 0.4 release, but there is a bug in handling LR schedule (fixed in main branch).
